Drop -Y flags that change language semantics

adriaanm opened the pull request:
removes:

* `-Yno-adapted-args`
* `-Xstrict-inference`
* `-Xfull-lubs`
* `-Yoverride-objects`
* `-Yoverride-vars`
* `-Yinfer-argument-types`
* `-Yvirtpatmat`

See https://github.com/scala/scala-dev/issues/430

This is the first cut. Open for discussion.

adriaanm committed:
Drop flags that change language semantics

To reduce the many possible combination of these flags,
the only supported mechanism that affects the outcome of
type checking is `-Xsource`, which instructs the compiler
to mimic the given version as much as possible. We only
support this for one prior and one subsequent version
(e.g., 2.13 recognizes 2.12 and 2.14 source levels).

The following flags have been dropped:
  - `-Xfull-lubs`
  - `-Xstrict-inference`
  - `-Yinfer-argument-types`
  - `-Yno-adapted-args`
  - `-Yoverride-objects`
  - `-Yoverride-vars`
  - `-Yvirtpatmat`

We're exceptionally dropping a few obscure `-X` flags
without prior deprecation. `-Y` flags are always at our whim.

adriaanm commented:
The last commit failed, largely because of check file changes: https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/741/testReport/ (it's sooooo nice to have a page like this to link to!)

My understanding is that we should put the FINAL bit back on objects, since its removal was a half-finished change to support the -Yoverride-objects flag.

eed3si9n commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -203,7 +199,6 @@ trait ScalaSettings extends AbsScalaSettings
   val Ynogenericsig   = BooleanSetting    ("-Yno-generic-signatures", "Suppress generation of generic signatures for Java.")
   val noimports       = BooleanSetting    ("-Yno-imports", "Compile without importing scala.*, java.lang.*, or Predef.")
   val nopredef        = BooleanSetting    ("-Yno-predef", "Compile without importing Predef.")
-  val noAdaptedArgs   = BooleanSetting    ("-Yno-adapted-args", "Do not adapt an argument list (either by inserting () or creating a tuple) to match the receiver.")
```
Kind of sad to see this one go. I only learned about it recently, but I thought it was a good cop.

adriaanm replied:
This PR is part provocation to hear any passionate defenses of the defendants :-)

milessabin replied:
I'm quite happy to see that one go. As Scala sharp edges go it's one of the least harmful (you're unlikely to silently miscompile because of it) and most useful (see various bits of arity-polymorphic sugar in shapeless and thereabouts).

eed3si9n replied:
Here's an easy example of things compiling:

```scala
scala> case class W[A](elem: A)(implicit ordering: Ordering[A])
defined class W

scala> W(1, 2, 3, 4)
res0: W[(Int, Int, Int, Int)] = W((1,2,3,4))
```

And to quote @lrytz in https://github.com/scala/bug/issues/8342:

> Auto-tupling is not mentioned at all in the spec.

Here's another [voice from the wild](https://twitter.com/djspiewak/status/430758303119863808):

<img width="638" alt="AGGGHGHGHGHGHGHGHGHGHGHGHGH!
Scala auto-tupling. Why on earth does this “feature” even exist?" src="https://user-images.githubusercontent.com/184683/38578590-5d5d1fb4-3cd2-11e8-80dc-2dbdd5ca6fad.png">

To which @tpolecat responds "`-Yno-adapted-args`."

I am not so much for or against the removal of `-Yno-adapted-args` per se, but I'd request a more principled solution toward this general problem. Adaptations.scala is already riddled with flags (`settings.future`, `settings.noAdaptedArgs`, `settings.warnAdaptedArgs`).

### option 1

Here's a proposal. Remove zero-arg eta expansion in 2.13 (or under -Xsource:2.14), and also remove auto-tupling under -Xsource:2.14 if Dotty can go along with it.

adriaanm replied:
> Here's a proposal. Remove zero-arg eta expansion in 2.13

Yup: https://github.com/scala/scala/pull/6475

> remove auto-tupling

Would be nice...

tpolecat replied:
I find that auto-tupling hinders readability, and at least for me it ends up being applied accidentally often enough that I prefer to turn it off. The 22-element limit makes it unusable for arity-abstraction so I don't really buy that argument. So I'm 👎for removing the flag and 👍for removing the feature.

adriaanm replied:
Been saying that for 4 years now ;-) https://github.com/scala/bug/issues/3583#issuecomment-292383595

sjrd replied:
One data point: a very nasty issue (https://github.com/scala-js/scala-js/issues/3154, fix in https://github.com/scala-js/scala-js/commit/9428b684197cea946467f9caeaefab616546d356) that could have been avoided if auto-tupling did not exist.

So yup, I'm not a fan of auto-tupling ...

adriaanm replied:
It's more restricted in dotty though: http://dotty.epfl.ch/docs/reference/auto-parameter-tupling.html. Looks more acceptable, though I wouldn't mind just spelling it out to avoid the gotcha in other places.

adriaanm replied:
In any case, our guiding principle is alignment with Dotty, so, looks like we can at least deprecate auto-tupling in most cases in 2.13 (drop it in 2.14), and implement the same approach as dotty by 2.14.

joroKr21 replied:
Dotty still has the regular auto tupling which I don't care much about. But the function parameter tupling is something different that doesn't exist in scalac and I find it quite useful.

adriaanm replied:
Ah, thanks for the correction. Wasn't aware they still had auto-tupling :-( I would really like to drop that, while I agree the one for function literals seems useful.

eed3si9n replied:
In an effort for reconciliation, I've opened a thread here https://contributors.scala-lang.org/t/lets-drop-auto-tupling/1799, please comment and subscribe.

milessabin replied:
@eed3si9n sure, you can construct examples of that sort. It's a lot harder to find examples where that's a problem in actual practice. They exist, but they're rare.

puffnfresh commented:
`-Yno-adapted-args` is very important to the projects I work on in Atlassian.

`-Xfuture` also changes semantics (e.g. the `withFilter` fallback) in a very useful way. Will it also be removed?

adriaanm commented:
I'm not for auto-tupling (see above), but I'm against flags that fragment the language. They cause confusion, reduce test coverage, complicate code,... The intent is to keep only `-Xsource`, to have the compiler do its best to emulate one version in the past or future (which subsumes `-Xfuture`).

I'd like to understand both your use cases better so we can see about supporting them by default. If we can convince the dotty folks to drop auto-tupling, I'd be happy to deprecate in 2.13 and remove in 2.14. `withFilter` is already the only supported method for desugaring `if` in a `for` in 2.12 -- the fallback to `filter` has been removed (https://github.com/scala/scala/pull/5252).

adriaanm commented:
I've dropped the commit about restoring the acc_final flag for nested objects and rebased (will create a ticket). Would like to consider merging this for M4.

dwijnand commented:
Does it makes sense, for now, to retain the auto-tupling flag, given there's at least some interest by Martin in dropping it in Scala 3?

SethTisue commented:
for the record, I'm confident `-Xstrict-inference` can go, I did due diligence on it over at #6593 before Adriaan pointed out it was done here already.

adriaanm commented:
I note your hesitation on auto-tupling, @dwijnand. I'd like for us to accept slight regressions in the process of cleaning this up. I've logged this as https://github.com/scala/scala-dev/issues/496.

som-snytt commented:
Now all my PRs with new `-Y` options have merge conflicts.

I only just noticed that the letter `Y` represents a fork and is really the best symbol for options that change semantics.

Similarly, `-X` should be reserved for chiasmus options that invert the meaning of your code.

SethTisue commented:
I'm (still) confused why `-Yno-adapted-args` got the axe. I've opened a separate ticket on it; let's discuss there: https://github.com/scala/bug/issues/11110

som-snytt commented:
They're all adapted args, @SethTisue in the sense of "good dogs". Yes, that's pretty funny and I deserve some sort of community award for it. Maybe when dotty becomes boring Scala 3, you could appropriate "Dotty" for the name of the award.
